File: client/python/example/quick/testEvent.py
import grpc
import event_pb2

from example.event.event import EventSvc

MAX_NUM_OF_LOG = 16
MAX_NUM_OF_IMAGE_LOG = 2
LOG_IMAGE_NAME = './image_log.jpg'
EVENT_QUEUE_SIZE = 16
QUEUE_SIZE = 16
MAX_REALTIME_LOG = 2

def testRealtimeLog(eventSvc, deviceID):
  try:
    eventSvc.enableMonitoring(deviceID)

    print(f'\n>>> Generate {MAX_REALTIME_LOG} events...', flush=True)
    eventCh = eventSvc.subscribe(QUEUE_SIZE)
    eventCount = 0

    for event in eventCh:
      print(f'Event: {event}', flush=True)
      eventCount += 1
      if eventCount >= MAX_REALTIME_LOG:
        eventCh.cancel()

    eventSvc.disableMonitoring(deviceID)

  except grpc.RpcError as e:
    if e.code() == grpc.StatusCode.CANCELLED:
      print('Subscription is cancelled', flush=True)    
    else:
      print(f'Cannot complete the realtime event test: {e}', flush=True)    
    

def testEvent(eventSvc, deviceID):
  try:
    events = eventSvc.getLog(deviceID, 0, MAX_NUM_OF_LOG)
    print(f'Events: \n{events}', flush=True)

    imageEvents = eventSvc.getImageLog(deviceID, 0, MAX_NUM_OF_IMAGE_LOG)
    print(f'Num of image events: {len(imageEvents)}', flush=True)

    if len(imageEvents) > 0:
      f = open(LOG_IMAGE_NAME, 'wb')
      f.write(imageEvents[0].JPGImage)
      f.close()

    testRealtimeLog(eventSvc, deviceID)

  except grpc.RpcError as e:
    print(f'Cannot complete the event test: {e}')    

File: client/node/example/quick/testEvent.js
const event = require('../event');
const fs = require('fs');

const MAX_NUM_OF_LOG = 16;
const MAX_NUM_OF_IMAGE_LOG = 2;
const LOG_IMAGE_NAME = "./image_log.jpg";
const EVENT_QUEUE_SIZE = 16;

function testRealtime(devID) {
  var sub;

  return event.enableMonitoring(devID)
  .then(() => {
    sub = event.subscribe(EVENT_QUEUE_SIZE);

    sub.on('data', (event) => {
      console.log('Event: ', event.toObject());
    });

    sub.on('end', () => {
      console.log('Subscription is finished');
    });

    sub.on('error', (err) => {
      console.log('Subscription error: ', err);
    })

    console.log('>>> Generate events for 10 seconds');

    return new Promise((resolve) => setTimeout(resolve, 10000));
  })
  .then(() => {
    sub.cancel();
    
    return event.disableMonitoring(devID);
  })
  .catch((err) => {
    console.error('Cannot finish the realtime event test: ', err);
    throw err;
  });
}

function test(devID) {
  return event.getLog(devID, 0, MAX_NUM_OF_LOG)
  .then((events) => {
    console.log('Events: ', events);

    return event.getImageLog(devID, 0, MAX_NUM_OF_IMAGE_LOG)
  })
  .then((imageEvents) => {
    console.log('Number of image events: ', imageEvents.length);

    if(imageEvents.length > 0) {
      let buf = new Buffer(imageEvents[0].jpgimage, 'base64');

      fs.writeFileSync(LOG_IMAGE_NAME, buf);
    } 

    return testRealtime(devID);
  })
  .catch((err) => {
    console.error('Cannot finish the event test: ', err);
    throw err;
  });
}

module.exports.test = test;